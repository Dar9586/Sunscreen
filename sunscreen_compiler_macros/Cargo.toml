[package]
name = "sunscreen_compiler_macros"
version = "0.4.0"
edition = "2021"

[lib]
proc-macro = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
proc-macro2 = "1.0.32"
quote = "1.0.10"
syn = { version = "1.0.81", features = ["derive", "full", "fold"] }

[dev-dependencies]
serde_json = "1.0.72"
