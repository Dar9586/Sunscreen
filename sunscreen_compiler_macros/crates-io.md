This crate is a component of the [Sunscreen compiler](https://crates.io/sunscreen).