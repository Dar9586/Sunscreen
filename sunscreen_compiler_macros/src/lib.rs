#![deny(missing_docs)]
#![deny(rustdoc::broken_intra_doc_links)]
#![recursion_limit = "128"]

//! This crate contains macros to support the sunscreen compiler.

extern crate proc_macro;

mod circuit;
mod decrypt;
mod error;
mod internals;
mod type_name;

#[proc_macro_derive(TypeName)]
/**
 * Allows you to `#[derive(Typename)]`.
 */
pub fn derive_typename(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    type_name::derive_typename(input)
}

#[proc_macro_attribute]
/**
 * Specifies a function to be a circuit. A circuit has any number of inputs that impl the
 * `FheType` trait and returns either a single type implementing `FheType` or a tuple of
 * types implementing `FheType`.
 *
 * This function gets run by the compiler to build up the circuit you specify and does not
 * directly or eagerly perform homomorphic operations.
 *
 * # Parameters
 * * `scheme` (required): Designates the scheme this circuit uses. Today, this must be `"bfv"`.
 *
 * # Examples
 * ```rust
 * # use sunscreen_compiler::{circuit, types::Unsigned, Params, Context};
 *
 * #[circuit(scheme = "bfv")]
 * fn multiply_add(a: Unsigned, b: Unsigned, c: Unsigned) -> Unsigned {
 *   a * b + c
 * }
 * ```
 *
 * ```rust
 * # use sunscreen_compiler::{circuit, types::Unsigned, Params, Context};
 *
 * #[circuit(scheme = "bfv")]
 * fn multi_out(a: Unsigned, b: Unsigned, c: Unsigned) -> (Unsigned, Unsigned) {
 *   (a + b, b + c)
 * }
 * ```
 */
pub fn circuit(
    metadata: proc_macro::TokenStream,
    input: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    circuit::circuit_impl(metadata, input)
}

#[proc_macro]
/**
 * Decrypts an output parameter set using the given runtime. The first argument
 * to this macro is an identifier to a runtime. The second argument is the identifier
 * of the return bundle to decrypt. 3rd-Nth arguments are the expected return types
 * from the circuit, in order. The macro returns a `Result<sunscreen_compiler::Error>`.
 * 
 * # Remarks
 * This macro validates the given types against the circuit's return interface
 * for correctness, then decrypts each item. If successful, this macro returns
 * an Ok(T) where T is:
 * * The unit type `()` if the circuit returned nothing.
 * * The single argument matching the lone type parameter
 * if the circuit returns one argument.
 * * A tuple of composed of the types passed to the macro if the circuit returns
 * more than one argument.
 * 
 * The types passed in arguments 3-N must exactly match those in the return interface
 * of the circuit. Circuits that return nothing, while useless, are legal. In this case,
 * you should only pass the first two arguments. In the event of failure, this function
 * returns the underlying issue.
 */
pub fn decrypt(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    decrypt::decrypt_impl(input)
}