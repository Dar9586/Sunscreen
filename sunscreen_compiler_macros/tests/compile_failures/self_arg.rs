use sunscreen_compiler_macros::{fhe_program};

struct Foo {}

impl Foo {
    #[fhe_program(scheme = "bfv")]
    fn panic_circuit(&self) {
    }
}

fn main() {
    
}