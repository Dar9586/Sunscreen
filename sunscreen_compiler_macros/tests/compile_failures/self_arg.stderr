error: fhe_program must not take a reference to self
 --> tests/compile_failures/self_arg.rs:6:5
  |
6 |     #[fhe_program(scheme = "bfv")]
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the attribute macro `fhe_program` (in Nightly builds, run with -Z macro-backtrace for more info)
